Execute tests with clang sanitizers

Issue #389 could have been prevented if the UB sanitizer had been used during continuous integration. We shall add a Travis job for this if possible.

Required flags:
```
-g -O2 -fsanitize=address -fsanitize=undefined -fno-omit-frame-pointer
```

the build on travis is done:
https://travis-ci.org/Daniel599/json/jobs/186639548
I thought it would be green (was OK on my PC with gcc 5.4) but it still has LeakSanitizer errors,
the good news, the UB errors are gone 😄
how do I know what packages did travis install? I don't see it the logs

Still no idea how your clang build works, would like to look into that later,
I`ll try to investigate about the leaks.

**side note:** 
On my PC I got some warnings regarding shifts (using gcc 5.4) which I didn't see on travis (which is also weird o.0)

> In file included from src/unit-cbor.cpp:31:
> ../src/json.hpp:6244:60: warning: shift count >= width of type
>       [-Wshift-count-overflow]
>                 vec.push_back(static_cast<uint8_t>((number >> 070) & 0xff));
>                                                            ^  ~~~
> ../src/json.hpp:6724:21: note: in instantiation of function template
>       specialization 'nlohmann::basic_json<std::map, std::vector,
>       std::__cxx11::basic_string<char>, bool, long long, unsigned long long,
>       double, std::allocator>::add_to_vector<unsigned int>' requested here
>                     add_to_vector(v, 1, N);
>                     ^
> ../src/json.hpp:7622:9: note: in instantiation of member function
>       'nlohmann::basic_json<std::map, std::vector,
>       std::__cxx11::basic_string<char>, bool, long long, unsigned long long,
>       double, std::allocator>::to_cbor_internal' requested here
>         to_cbor_internal(j, result);
>         ^
> src/unit-cbor.cpp:44:39: note: in instantiation of member function
>       'nlohmann::basic_json<std::map, std::vector,
>       std::__cxx11::basic_string<char>, bool, long long, unsigned long long,
>       double, std::allocator>::to_cbor' requested here
>             const auto result = json::to_cbor(j);
>                                       ^`
> 

I don't think it's big deal since all your uses are with constant a mount of bytes and their size matches,
but if people use with your code -Werror, that might get in their way 
I can look into that as well if you think it's needed

--- a/.travis.yml
+++ b/.travis.yml
@@ -42,14 +42,20 @@ matrix:
       - make check TEST_PREFIX="valgrind --error-exitcode=1 --leak-check=full " TEST_PATTERN=""
 
   # cLang sanitizer
-
-  #- os: linux
-  #  env:
-  #    - LLVM_VERSION=3.8.1
-  #    - SPECIAL=sanitizer
-  #  compiler: clang
-  #  before_script:
-  #    - make clang_sanitize
+  # note: sadly clang's libc++ has errors when running with sanitize,
+  # so we use clang with gcc's libstdc++ which doesn't give those error.
+  # that's why we need to install g++-6 to get the lastest version
+  - os: linux
+    env:
+      - LLVM_VERSION=3.8.1
+      - SPECIAL=sanitizer
+    addons:
+      apt:
+        sources: ['ubuntu-toolchain-r-test']
+        packages: g++-6
+    compiler: clang
+    before_script:
+      - make clang_sanitize
 
   # cppcheck
 

--- a/Makefile
+++ b/Makefile
@@ -92,8 +92,9 @@ fuzzing-stop:
 cppcheck:
 	cppcheck --enable=warning --inconclusive --force --std=c++11 src/json.hpp --error-exitcode=1
 
+# run clang sanitize (we are overrding the CXXFLAGS provided by travis in order to use gcc's libstdc++)
 clang_sanitize: clean
-	CXX=clang++ CXXFLAGS="-g -O2 -fsanitize=address -fsanitize=undefined -fno-omit-frame-pointer" $(MAKE)
+	CXX=clang++ CXXFLAGS="-g -O2 -fsanitize=address -fsanitize=undefined -fno-omit-frame-pointer" $(MAKE) check -C test
 
 
 ##########################################################################

--- a/test/src/unit-allocator.cpp
+++ b/test/src/unit-allocator.cpp
@@ -111,6 +111,16 @@ struct my_allocator : std::allocator<T>
     }
 };
 
+// allows deletion of raw pointer, usually hold by json_value
+template<class T>
+void my_allocator_clean_up(T* p)
+{
+    assert(p != nullptr);
+    my_allocator<T> alloc;
+    alloc.destroy(p);
+    alloc.deallocate(p, 1);
+}
+
 TEST_CASE("controlled bad_alloc")
 {
     // create JSON type using the throwing allocator
@@ -131,7 +141,8 @@ TEST_CASE("controlled bad_alloc")
             {
                 next_construct_fails = false;
                 auto t = my_json::value_t::object;
-                CHECK_NOTHROW(my_json::json_value j(t));
+                auto clean_up = [](my_json::json_value& j){ my_allocator_clean_up(j.object); };
+                CHECK_NOTHROW(my_json::json_value j(t); clean_up(j));
                 next_construct_fails = true;
                 CHECK_THROWS_AS(my_json::json_value j(t), std::bad_alloc);
                 next_construct_fails = false;
@@ -140,7 +151,8 @@ TEST_CASE("controlled bad_alloc")
             {
                 next_construct_fails = false;
                 auto t = my_json::value_t::array;
-                CHECK_NOTHROW(my_json::json_value j(t));
+                auto clean_up = [](my_json::json_value& j){ my_allocator_clean_up(j.array); };
+                CHECK_NOTHROW(my_json::json_value j(t); clean_up(j));
                 next_construct_fails = true;
                 CHECK_THROWS_AS(my_json::json_value j(t), std::bad_alloc);
                 next_construct_fails = false;
@@ -149,7 +161,8 @@ TEST_CASE("controlled bad_alloc")
             {
                 next_construct_fails = false;
                 auto t = my_json::value_t::string;
-                CHECK_NOTHROW(my_json::json_value j(t));
+                auto clean_up = [](my_json::json_value& j){ my_allocator_clean_up(j.string); };
+                CHECK_NOTHROW(my_json::json_value j(t); clean_up(j));
                 next_construct_fails = true;
                 CHECK_THROWS_AS(my_json::json_value j(t), std::bad_alloc);
                 next_construct_fails = false;
@@ -160,7 +173,8 @@ TEST_CASE("controlled bad_alloc")
         {
             next_construct_fails = false;
             my_json::string_t v("foo");
-            CHECK_NOTHROW(my_json::json_value j(v));
+            auto clean_up = [](my_json::json_value& j){ my_allocator_clean_up(j.string); };
+            CHECK_NOTHROW(my_json::json_value j(v); clean_up(j));
             next_construct_fails = true;
             CHECK_THROWS_AS(my_json::json_value j(v), std::bad_alloc);
             next_construct_fails = false;

